### ISSUE TITLE
RFC: Moving to a real graph DB?

### ISSUE THREAD

**Issue by jethrokuan** [jethrokuan/org-roam#199]
### Brief Abstract
Considering moving to a real graph database e.g. https://dgraph.io/

### Long Description
Right now, the caches are stored in Emacs-lisp hash tables. This makes several queries quite inconvenient, and some, inefficient. I think slapping a layer of data abstraction on top would be very helpful for maintainability in the future.

### Proposed Implementation (if any)
Evaluate:
- [Dgraph](https://dgraph.io/)
- [RedisGraph](https://oss.redislabs.com/redisgraph/)

### Please check the following:
- [ ] No similar feature requests

**Comment by jethrokuan** [comment]
Actually a simple sqlite database would do:

```
Update table:
| id | file   | last_modified |
|----+--------+---------------|
|  1 | f1.org | timestamp     |

With index on file

Content table:
| id | type     | from   | to     | val                  |
|----+----------+--------+--------+----------------------|
|  1 | link     | f1.org | f2.org | elisp serialized obj |
|  2 | citation | f2.org | bibkey | elisp serialized obj |

With index on from, to

Refs table:
| id | ref                | file   |
|----+--------------------+--------|
|  1 | https://google.com | f1.org |

With index on ref, file
```

May index generously because I don't think size will be a big issue, but at the same time, the tables are probably going to be really small. This may be slower than having them in memory in Emacs, but there are several benefits:

1. No need to rebuild from scratch, as data is serialized to disk
2. Use much less memory, hence less GC within Emacs
3. SQLite database can be read and used elsewhere

**Comment by alanz** [comment]
Would this allow concurrent access from multiple separate emacs instances?

**Comment by jethrokuan** [comment]
> Would this allow concurrent access from multiple separate emacs instances?

Theoretically yes. I'm not sure if the sqlite binary needs to be compiled differently though.

**Comment by jethrokuan** [comment]
Also we can locate the sqlite db at the root of the org-roam-directory, and supporting multiple org-roam instances becomes trivial.

**Comment by joefromct** [comment]
> This may be slower than having them in memory in Emacs, but there are several benefits.

You could still offer the flexibility to keep in memory [with sqlite ](https://www.sqlite.org/inmemorydb.html), although not exactly emacs memory.  

Not sure how the performance would compare, but this would offer the flexibility of having in-memory and still having segregation of duties which might be nice. And the user could change their persistence choices easily.  

thanks for all your work on this.

**Comment by jethrokuan** [comment]
Emacsql is making this really easy, most of the work is already done in https://github.com/jethrokuan/org-roam/pull/200. But it'll take a while to bring it back to feature parity I think. I'm not sure if emacsql's db connection is an in memory one.

**Comment by jethrokuan** [comment]
closed with #200
